### PR TITLE
Circuit cutting: Add JIT compilation tests

### PR DESCRIPTION
**Pull request by trbromley**
**Context:**

Adds tests for the `cut_circuit` transform that ensures JIT compilation with all of the JIT-supported interfaces.

### CODE REVIEW

File: tests/transforms/test_qcut.py [hunk 1]
```diff
@@ -1908,6 +1908,52 @@ def circuit(x):
...
+            return qml.expval(qml.PauliZ(wires=[0]))
+
+        x = torch.tensor(0.531, requires_grad=True)
+
+        # Note that the jit.trace ends up calling qcut_processing_fn multiple times, so below we
+        # delay introducing the spy until afterwards and then ensure that qcut_processing_fn is
+        # not called again.
+        cut_circuit_jit = torch.jit.trace(
+            qcut.cut_circuit(circuit, use_opt_einsum=use_opt_einsum), x
+        )
```
Comment (josh146) [inline tests/transforms/test_qcut.py:1938]: @trbromley very very minor but I'm not sure `torch.jit.trace` is actually JITing. I think `torch.jit.script` is the decorator that actually performs compilation (but also don't think it works with PL).

Having said that, tracing is still good to test and still faster than non-tracing.

**Comment by trbromley** [comment]
[sc-14724]

File: tests/transforms/test_qcut.py [hunk 2]
```diff
@@ -2105,6 +2105,140 @@ def circuit(x):
...
+
+        spy = mocker.spy(qcut, "qcut_processing_fn")
+        res = cut_circuit_jit(x)
+
+        assert spy.call_count == 0
```
Comment (anthayes92) [inline tests/transforms/test_qcut.py:2140]: could use [`assert_not_called()`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_not_called) ?

File: tests/transforms/test_qcut.py (hunk 1 above)
Comment (trbromley) [reply tests/transforms/test_qcut.py]: Thanks @josh146! I reworded here: 9bf8d0de7817ff0f1d8048f3cfd8bdf4b7d8b522. Couldn't get scripting to work!

File: tests/transforms/test_qcut.py (hunk 2 above)
Comment (trbromley) [reply tests/transforms/test_qcut.py]: Good spot! Fixed here: 79f9331bb94a3fe7d5b9526a5ec71e59c0157d3e

File: tests/transforms/test_qcut.py [hunk 3]
```diff
@@ -2091,6 +2091,203 @@ def circuit(x):
...
+            res_expected = circuit(x)
+
+            assert np.isclose(res, res_expected)
+
+            grad = jax.grad(cut_circuit_jit)(x)
+            grad_expected = jax.grad(circuit)(x)
+
+            assert np.isclose(grad, grad_expected)
+
+        assert spy.call_count == 4
```
Comment (anthayes92) [inline tests/transforms/test_qcut.py:2289]: how has this increased by 2 from the counts before the loop? Once the code has been complied should `qcut_processing_fn` be called at all?

**Review by anthayes92 — APPROVED** [review]
Thanks @trbromley! Looks good to go now!

Out of interest, I saw some trouble earlier on in this PR, what was the difficulty there?

File: tests/transforms/test_qcut.py (hunk 3 above)
Comment (trbromley) [reply tests/transforms/test_qcut.py:2289]: I think JIT/tracing is quite dependent on the interface. For JAX, I think it compiles once on the forward pass and once for the gradient. For some reason it does this again in the `for` loop, so we have `2 x 2` executions.

**Comment by trbromley** [comment]
> Out of interest, I saw some trouble earlier on in this PR, what was the difficulty there?

What was the trouble? [This](https://github.com/PennyLaneAI/pennylane/pull/2244#discussion_r815918526) comment? We are currently using `torch.jit.trace` rather than `torch.jit.script`. Scripting is more restricted and incompatible with `cut_circuit` and possibly parts of PennyLane more generally.
